File: Data Science/python/chicks.py
import random

unpecked = []
iters = 3000

for _ in range(iters):

    chicks = [0 for i in range(100)]

    for i in range(100):
        l_r = random.choice([-1, 1])

        if i + l_r > 99:
            i = i - 100
        elif i + l_r < 0:
            i = i + 100

        chicks[i + l_r] = 1

    up = 0

    for chick in chicks:
        if chick == 0:
            up += 1

    unpecked.append(up)

total_up = 0

for i in unpecked:
    total_up += i

average = (total_up / 100) / len(unpecked) * 100

print(f"Average pecked over {iters} iterations: {average}%")

File: Data Science/python/chicks.c
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>

#include <time.h>

typedef uint32_t chick_t;

void chick_peck(uint32_t* unpecked) {
    // Set all of the chickens to being unpecked
    chick_t* chickens = calloc(sizeof(chick_t), sizeof(chick_t) * 100);

    for (int i = 0; i < 100; i++) {
        int choice = (rand() % 2);
        int side = 0;

        int idx = i;

        switch (choice) {
            case 0:
                side = -1;
                break;
            case 1:
                side = 1;
                break;
        }

        idx = idx + side;

        if (idx > 99) {
            idx = idx - 100;
        } else if (idx < 0) {
            idx = idx + 100;
        }

        chickens[idx + side] = 1;

    }

    uint32_t up = 0;

    for (int i = 0; i < 100; i++) {
        if (chickens[i] == 0) {
            up++;
        }
    }

    // We dont want a memory leak :)
    free(chickens);

    *unpecked += up;

}

int main() {

    time_t t;

    srand((unsigned) time(&t));

    int iters = 100;

    uint32_t total = 0;

    for (int i = 0; i < iters; i++) {
        chick_peck(&total);
    }

    double percentage = (total / 100) / iters * 100;

    printf("Average pecked over %d iterations: %f\n", iters, percentage);


    return 0;
}
